rubix-core/src/main/java/com/qubole/rubix/core/CachedReadRequestChain.java

```diff
@@ -48,19 +75,28 @@ public Integer call()
 
         checkState(isLocked, "Trying to execute Chain without locking");
         for (ReadRequest readRequest : readRequests) {
-            fileToRead.seek(readRequest.getActualReadStart());
             int nread = 0;
+            int leftToRead = readRequest.getActualReadLength();
+            log.debug(String.format("Processing readrequest %d-%d, length %d", readRequest.actualReadStart, readRequest.actualReadEnd, leftToRead));
             while (nread < readRequest.getActualReadLength()) {
-                int nbytes = fileToRead.read(readRequest.getDestBuffer(), readRequest.getDestBufferOffset() + nread, readRequest.getActualReadLength() - nread);
-                log.debug(String.format("CachedFileRead copied data from %d of length %d at buffer offset %d",
-                        readRequest.getActualReadStart(),
-                        readRequest.getActualReadLength() - nread,
-                        readRequest.getDestBufferOffset() + nread));
-                if (nbytes < 0) {
+                int readInThisCycle = Math.min(leftToRead, directBuffer.capacity());
+
+                directBuffer.position(0);
+                int nbytes = fileChannel.read(directBuffer, readRequest.getActualReadStart() + nread);
```

Comment:
We know how many bytes to expect in total but not in a single iteration. This acts as offset for read in next iteration.